### PR TITLE
Allow external constrains in generated code

### PR DESCRIPTION
**Pull request by dmdashenkov**
As for now, generated Java validation code ignores the `(costraint_for)` options, which define external constraints for a given message field.

This PR adds support for external constraints in generated code.

By their nature, external constraints cannot be discovered during code generation. That's why, instead of converting them into validation code, as we do with the usual constraints, we assemble and check them at runtime. The generated code merely calls those checks if any external constraints are to be found.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: base/src/main/java/io/spine/validate/ExternalConstraints.java [hunk 1]
```diff
@@ -80,6 +81,30 @@ private ExternalConstraints(ImmutableSet<ExternalMessageConstraint> constraints)
         return Holder.instance.constraints;
     }
 
+    /**
+     * Checks if there is in external constraint targeting the field with the given name in
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/ExternalConstraints.java]: `an`.

File: base/src/main/java/io/spine/validate/ExternalConstraints.java [hunk 2]
```diff
@@ -80,6 +81,30 @@ private ExternalConstraints(ImmutableSet<ExternalMessageConstraint> constraints)
...
+     * {@code definedFor(Customer.getDescriptor(), "name")} will return {@code true}.
+     *
+     * @param containerType
+     *         the type which contains the target field
+     * @param fieldName
+     *         the target field
+     * @return {@code true} if there is an external constraint targeting the field,
+     *         {@code false} otherwise
+     */
+    public static boolean definedFor(Descriptor containerType, String fieldName) {
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/ExternalConstraints.java]: I would prepend `is`.

File: base/src/main/java/io/spine/validate/ExternalConstraints.java [hunk 3]
```diff
@@ -80,6 +81,30 @@ private ExternalConstraints(ImmutableSet<ExternalMessageConstraint> constraints)
...
+        for (ExternalMessageConstraint constraint : all()) {
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/ExternalConstraints.java]: This one looks `anyMatch()`.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/ClassMember.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import com.squareup.javapoet.TypeSpec;
+
+/**
+ * A member of a Java class.
+ *
+ * <p>The member can be {@linkplain #attachTo attached} to a type constructed with Java Poet.
+ */
+public interface ClassMember {
+
+    void attachTo(TypeSpec.Builder type);
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/ClassMember.java]: Please document.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import io.spine.validate.ExternalConstraints;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static javax.lang.model.element.Modifier.FINAL;
+import static javax.lang.model.element.Modifier.PRIVATE;
+import static javax.lang.model.element.Modifier.STATIC;
+
+/**
+ * A boolean flag in the generated code which signifies whether or not a field has externally
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java]: `or not` looks redundant to me. `whether` is enough.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ * defined constraints.
+ *
+ * <p>Such a flag is only generated for message fields marked with {@code (validate) = true}. Other
+ * fields cannot be target for external constraints.
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java]: `a target`, `targets`, `targeted`?

File: tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    BooleanExpression value() {
+        return field;
+    }
+
+    /**
+     * Obtains this flag as a {@link ClassMember}.
+     */
+    ClassMember asClassMember() {
+        ClassName containingTypeName = ClassName.bestGuess(declaration.declaringType()
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/ExternalConstraintFlag.java]: As the lookup seems expensive, I would cache the result. Or compute it right away in the ctor.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/Field.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+package io.spine.tools.validate;
+
+import com.google.common.base.Objects;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.TypeSpec;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A field to be attached to a Java class.
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/Field.java]: Please say a bit more on that "attachment". It's quite difficult for an out-of-context comprehension.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/Method.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+package io.spine.tools.validate;
+
+import com.google.common.base.Objects;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.TypeSpec;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A method to be attached to a Java class.
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/Method.java]: Same as above.

File: tools/validation-generator/src/main/java/io/spine/tools/validate/code/ConditionalStatement.java [hunk 10]
```diff
@@ -61,6 +61,41 @@
...
+    }
+
+    /**
+     * Adds the {@code else} branch to this statement.
+     *
+     * <p>Completes this statement. If the statement is already complete, throws
+     * an {@link IllegalStateException}.
+     *
+     * @param negativeBranch
+     *         the alternative branch
```
Comment (armiol) [inline tools/validation-generator/src/main/java/io/spine/tools/validate/code/ConditionalStatement.java]: Consider calling the parameter so.

File: base/src/main/java/io/spine/validate/ExternalConstraints.java (hunk 3 above)
Comment (dmdashenkov) [reply base/src/main/java/io/spine/validate/ExternalConstraints.java]: As discussed, we use a loop here due to performance considerations.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Pull request merged**